src/daemon/pcsc_context.cpp

```diff
@@ -0,0 +1,379 @@
+
+            if (states[NOTIFICATION_STATE_INDEX].dwEventState & SCARD_STATE_CHANGED) {
+                // Looks like \\?PnP?\Notification can miss readers in pcsc-lite. Number of readers
+                // are counted at start of SCardGetStatusChange() and \\?PnP?\Notification is only
+                // set to "changed" if the number of readers change. If a reader is added between
+                // SCardListReaders() and SCardGetStatusChange(), it will not be noticed. Sleeping a
+                // bit makes it less likely that a reader is not detected in next SCardListReaders()
+                // call. Have observed this behavior with the ACR1252 that exposes two readers. It
+                // is very often both of them are not detected without this sleep.
+                std::this_thread::sleep_for(std::chrono::milliseconds(100));
+                reader_names = list_readers(context);
+                states = initial_states(reader_names);
+            }
+        }
+
+        ret = SCardReleaseContext(context);
+        if (ret != SCARD_S_SUCCESS)
+            g_warning("Failed to release PC/SC context: %s", pcsc_stringify_error(ret));
+    }
+
+    void PCSCContext::thread_join()
+    {
+        {
+            std::unique_lock<std::mutex> lock(run_status_.mutex);
+
+            run_status_.stop = true;
+
+            // SCardCancel() may not be noticed so loop until it is. Condition to wake up
+            // as quickly as possible when it is done and to retry if it is not. See
+            // https://salsa.debian.org/rousseau/PCSC/issues/16 for upstream bug.
+            while (run_status_.cancellable) {
+                SCardCancel(context_);
+                run_status_.cancelled_condition.wait_for(lock, std::chrono::milliseconds(50));
+            }
+        }
+
+        thread_.join();
+    }
+
+    void PCSCContext::check_states_after_get_status_change(SCARDCONTEXT context,
+                                                           std::vector<SCARD_READERSTATE> &states)
+    {
+        for (std::size_t i = 0; i < states.size(); i++) {
+            if (i == NOTIFICATION_STATE_INDEX)
+                continue;
+
+            SCARD_READERSTATE &state = states[i];
+
+            if (state.dwEventState & SCARD_STATE_CHANGED) {
+                if ((state.dwCurrentState & SCARD_STATE_EMPTY) &&
+                    (state.dwEventState & SCARD_STATE_PRESENT)) {
+                    card_present(context, state);
+                }
+            }
+
+            state.dwCurrentState = state.dwEventState;
+        }
+    }
+
+    void PCSCContext::card_present(SCARDCONTEXT context, const SCARD_READERSTATE &state)
+    {
+        if (uid_extract_) {
+            if (get_data_uid_supported(state)) {
+                auto uid = transmit_get_data_uid(context, state.szReader);
+
+                if (uid)
+                    uid_queue_.push(ExtractedUID{std::move(*uid), state.szReader});
```

Comment:
Consider add missing compound-statement

Reply:
Pushed a fixup commit that makes new code follow AUTOSAR C++14 rule.